pkg/awsclient/elb_helper.go

```diff
@@ -0,0 +1,170 @@
+func (c *awsClient) removeListenersFromELB(elbName string) error {
+	i := &elb.DeleteLoadBalancerListenersInput{
+		LoadBalancerName:  aws.String(elbName),
+		LoadBalancerPorts: aws.Int64Slice([]int64{6443}),
+	}
+	_, err := c.DeleteLoadBalancerListeners(i)
+	return err
+}
+
+// addListenersToELB will add the +listeners+ to the specified ELB. This is
+// useful for when the "ext" ELB is to be publicly accessible. See also
+// removeListenersFromELB.
+// Note: This will likely always want to be given 6443/tcp -> 6443/tcp for
+// the kube-api
+func (c *awsClient) addListenersToELB(elbName string, listeners []*elb.Listener) error {
+	i := &elb.CreateLoadBalancerListenersInput{
+		Listeners:        listeners,
+		LoadBalancerName: aws.String(elbName),
+	}
+	_, err := c.CreateLoadBalancerListeners(i)
+	return err
+}
+
+// AddLoadBalancerInstances will attach +instanceIds+ to +elbName+
+// so that they begin to receive traffic. Note that this takes an amount of
+// time to return. This is also additive (but idempotent - TODO: Validate this).
+// Note that the recommended steps:
+// 1. stop the instance,
+// 2. deregister the instance,
+// 3. start the instance,
+// 4. and then register the instance.
+func (c *awsClient) AddLoadBalancerInstances(elbName string, instanceIds []string) error {
+	instances := make([]*elb.Instance, 0)
+	for _, instance := range instanceIds {
+		instances = append(instances, &elb.Instance{InstanceId: aws.String(instance)})
+	}
+	i := &elb.RegisterInstancesWithLoadBalancerInput{
+		Instances:        instances,
+		LoadBalancerName: aws.String(elbName),
+	}
+	_, err := c.RegisterInstancesWithLoadBalancer(i)
+	return err
+}
+
+// RemoveInstancesFromLoadBalancer removes +instanceIds+ from +elbName+, eg when an Node is deleted.
+func (c *awsClient) RemoveInstancesFromLoadBalancer(elbName string, instanceIds []string) error {
+	instances := make([]*elb.Instance, 0)
+	for _, instance := range instanceIds {
+		instances = append(instances, &elb.Instance{InstanceId: aws.String(instance)})
+	}
+	i := &elb.DeregisterInstancesFromLoadBalancerInput{
+		Instances:        instances,
+		LoadBalancerName: aws.String(elbName),
+	}
+	_, err := c.DeregisterInstancesFromLoadBalancer(i)
+	return err
+}
+
+// DoesELBExist checks for the existence of an ELB by name. If there's an AWS
+// error it is returned.
+func (c *awsClient) DoesELBExist(elbName string) (bool, *AWSLoadBalancer, error) {
+	i := &elb.DescribeLoadBalancersInput{
+		LoadBalancerNames: []*string{aws.String(elbName)},
+	}
+	res, err := c.DescribeLoadBalancers(i)
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case elb.ErrCodeAccessPointNotFoundException:
+				return false, &AWSLoadBalancer{}, nil
+			default:
+				return false, &AWSLoadBalancer{}, err
+			}
+		}
+	}
+	return true, &AWSLoadBalancer{ELBName: elbName, DNSName: *res.LoadBalancerDescriptions[0].DNSName, DNSZoneId: *res.LoadBalancerDescriptions[0].CanonicalHostedZoneNameID}, nil
```

Comment:
why not just return the elb in res from line 143 directly?

Reply:
I don't want the controller(s) to deal in cloud-specific objects where possible. Clearly here the `AWS` gives it away, but in the future, this may change name or type (eg to interface).